### DISCUSSION TITLE
Cannot extract EXIF Orientation tag on certain images

### DISCUSSION THREAD

**Discussion by winterstefan** [libvips/libvips#4073] [General]
## Bug report

Having two different `jpg` images with an EXIF `Orientation` tag present, for image `a.jpg` I can extract this tag where for image `b.jpg` vips doesn't return the orientation information. For these images, it's 100% reproducible. Other exif tools (e.g. `exiftool` or the Mac preview can extract the orientation for both images.

Unfortunately I cannot spot any technical differences between these two images, that my result in this behavior. Could you please take a look verifying the issue? Do you have any idea / hint? :-) 

**To Reproduce**
1. Download both images [a.jpg](https://github.com/user-attachments/assets/874d11dc-6a01-45ac-9015-18de61f1bb73) and [b.jpg](https://github.com/user-attachments/assets/6a203ff9-f6a5-442c-b94e-7596ad54b3f2)
2. Run `vipsheader -a ` for both images
3. Check output for `Orientation` tag

**Expected behavior**
Both commands display the exif orientation info

**Actual behavior**
For `a.jpg` I see an orientation, for `b.jpg` I don't.

```bash
$ vipsheader -a a.jpg | grep "Orientation"
exif-ifd0-Orientation: 6 (Rechts oben, Short, 1 components, 2 bytes)
$ vipsheader -a b.jpg | grep "Orientation"
```

Using e.g. `exiftool`, I see the following (also, when opening in Mac preview)
```bash
$ exiftool a.jpg | grep "Orientation"
Orientation                     : Rotate 90 CW
$ exiftool b.jpg | grep "Orientation"
Orientation                     : Rotate 270 CW
```

In addition, here's the output of using `autorot` directly:

```bash
# a.jpg - See width and height being swapped correctly
$ vips autorot a.jpg a-out.jpg && file a*
a-out.jpg: JPEG image data, Exif standard: [TIFF image data, big-endian, direntries=6, orientation=upper-left, xresolution=86, yresolution=94, resolutionunit=2], baseline, precision 8, 180x270, components 3
a.jpg:     JPEG image data, JFIF standard 1.01, resolution (DPI), density 72x72, segment length 16, Exif Standard: [TIFF image data, big-endian, direntries=5, orientation=upper-right, xresolution=74, yresolution=82, resolutionunit=2], baseline, precision 8, 270x180, components 3

# b.jpg - width and height stay identical
b-out.jpg: JPEG image data, Exif standard: [TIFF image data, big-endian, direntries=6, compression=JPEG (old), orientation=upper-left, xresolution=86, yresolution=94, resolutionunit=2], baseline, precision 8, 2653x1629, components 1
b.jpg:     JPEG image data, JFIF standard 1.01, resolution (DPI), density 72x72, segment length 16, Exif Standard: [TIFF image data, big-endian, direntries=15, compression=JPEG (old)], baseline, precision 8, 2653x1629, components 1
```

**Environment**
 - OS: MacOS Ventura 13.3.1
 - Vips: `vips-8.15.0`

**Additional context**
In general, I have a golang script using [govips](https://github.com/davidbyttow/govips). At one location I use `vipsheader -a` as binary to get all headers for an `io.Reader`. At another location I tried using `im.GetExif()` on a `vips.ImageRef`. Both failed providing the orientation for `a.jpg` (I also have a couple other images like `a.jpg` that doesn't work).

Being able to get the orientation tag is key for in order to rotate the image accordingly. At best, I'd use `im.Autorotate()`, but without it being able to extract the tag correctly it fails.

**Important** I also need to find a solution that works with streaming the image (because of large images being handled) 

- [a.jpg](https://github.com/user-attachments/assets/874d11dc-6a01-45ac-9015-18de61f1bb73)
- [b.jpg](https://github.com/user-attachments/assets/6a203ff9-f6a5-442c-b94e-7596ad54b3f2)

Thank you very much in advance; best regards

**Comment by jcupitt** [comment]
Hi @winterstefan,

I enabled debugging for the libvips EXIF loader and for `b.jpg` I see:

```
show_values: tags held in ExifData
tags in 'Motorola' byte order
Title|Value|Format|Size
- ifd 0
Compression|6|Short|2 bytes
- ifd 1
- ifd 2
- ifd 3
- ifd 4
```

So it looks like libexif is failing to parse anything useful from the 1035 bytes of EXIF in APP1.

With `a.jpg` I see:

```
show_values: tags held in ExifData
tags in 'Motorola' byte order
Title|Value|Format|Size
- ifd 0
Orientation|Right-top|Short|2 bytes
X-Resolution|72|Rational|8 bytes
Y-Resolution|72|Rational|8 bytes
Resolution Unit|Inch|Short|2 bytes
YCbCr Positioning|Centred|Short|2 bytes
- ifd 1
- ifd 2
- ifd 3
- ifd 4
```

Which looks more reasonable. I'll see if I can get any more information from libexif about this read failure.

**Comment by lovell** [comment]
The provided `b.jpg` is from a Samsung device and contains invalid, out-of-order EXIF metadata. It looks like the same underlying issue as https://github.com/lovell/sharp/issues/3947, due to the strict ordering requirement of libexif.

```
$ exiftool -validate -warning 6a203ff9-f6a5-442c-b94e-7596ad54b3f2 
Validate                        : 40 Warnings (14 minor)
Warning                         : Entries in IFD0 are out of order
```

```
$ exiftool -D 6a203ff9-f6a5-442c-b94e-7596ad54b3f2 
...
  259 Compression                     : JPEG (old-style)
  513 Other Image Start               : 868
  514 Other Image Length              : 32578
  271 Make                            : samsung
  274 Orientation                     : Rotate 270 CW
  306 Modify Date                     : 2024:03:07 14:58:49
...
```

**Comment by jcupitt** [comment]
Ah! Good sleuthing @lovell 

It sounds like there's not much libvips can do in this case.

**Comment by winterstefan** [comment]
**Oh, that indeed sounds reasonable! Thank you so much forgiving me a hint on this!**

Then, your last comment @lovell at https://github.com/lovell/sharp/issues/3947 is a great summary: For me as a consumer, it would be cool having libexif being able to also parse out-of-order tags (since I have a couple of customer images to process that seem to have this problem). But that explains why trying another golang package failed as well and thus is not a matter for libvips.

Thanks for checking that!

**Comment by lovell** [comment]
I'll move this to the discussions section should anyone fancy looking at the upstream libexif to see what might be possible. An alternative solution might be to tell all your friends to stop buying Samsung devices ;)

**Comment by lovell** [comment]
I was wrong about this relating to tag-ordering, but it is still ultimately a dodgy Samsung software problem.

It looks like some Samsung devices generate EXIF metadata with invalid thumbnail offsets and/or lengths.

libexif added protection against this when recursively handling EXIF directories via commit https://github.com/libexif/libexif/commit/cdf1e32cb71c22c3df5d806d74384b3189008a47 but this resulted in tag parsing aborting early, hence everything after the `Other Image Length` tag in the sample image being ignored.

I've proposed that libexif relaxes this logic slightly via https://github.com/libexif/libexif/pull/183

**Comment by tlmader** [comment]
@lovell We're seeing this issue with photos uploaded by some of our Samsung users. Photos are oriented correctly in their photo libraries before uploading, so this is difficult to prevent. I can't think of another way around it other than adding an in-app tool to rotate photos after uploading (not something I can take on right now).

If that proposal doesn't get merged, do you foresee a way around for Sharp users this without modifying libexif?

Validation with ExifTool:
```
exiftool -validate -warning -error nicks-ranger-seidel-1964-ash-gray-1734391131-1.jpeg
Validate                        : 44 Warnings (18 minor)
Warning                         : Entries in IFD0 are out of order
```
Result after Sharp: https://www.patinaproject.com/photos/nicks-ranger-seidel-1964-ash-gray/LmQKV1q

**Reply by lovell** [reply]
The prebuilt binaries for the future sharp v0.34.0 will be patched to include the proposed libexif change.

- https://github.com/lovell/sharp-libvips/commit/37f6402697c530fe4e0288d42dd8e20f4a076eb1
- https://github.com/libvips/build-win64-mxe/commit/3667699731bf51366cf9e6c7230a76b814431815

**Reply by tlmader** [reply]
Thank you, @lovell ! That's wonderful news.
